pkg/component/server/konnectivity.go

```diff
@@ -73,65 +74,99 @@ func (k *Konnectivity) Init() error {
 		return fmt.Errorf("failed to chown %s: %v", k.K0sVars.KonnectivitySocketDir, err)
 	}
 
-	// set default serverCount to 1
-	k.serverCount = kubernetes.MinLeaseHolders
+	// Buffered chan to send updates for the count of servers
+	k.serverCountChan = make(chan int, 1)
+
+	k.stopCtx, k.stopFunc = context.WithCancel(context.Background())
+
+	k.log = logrus.WithFields(logrus.Fields{"component": "konnectivity"})
+
 	return assets.Stage(k.K0sVars.BinDir, "konnectivity-server", constant.BinDirMode)
 }
 
 // Run ..
 func (k *Konnectivity) Run() error {
-	args := []string{
-		fmt.Sprintf("--uds-name=%s", filepath.Join(k.K0sVars.KonnectivitySocketDir, "konnectivity-server.sock")),
-		fmt.Sprintf("--cluster-cert=%s", filepath.Join(k.K0sVars.CertRootDir, "server.crt")),
-		fmt.Sprintf("--cluster-key=%s", filepath.Join(k.K0sVars.CertRootDir, "server.key")),
-		fmt.Sprintf("--kubeconfig=%s", k.K0sVars.KonnectivityKubeConfigPath),
-		"--mode=grpc",
-		"--server-port=0",
-		"--agent-port=8132",
-		"--admin-port=8133",
-		"--agent-namespace=kube-system",
-		"--agent-service-account=konnectivity-agent",
-		"--authentication-audience=system:konnectivity-server",
-		"--logtostderr=true",
-		"--stderrthreshold=1",
-		"-v=2",
-		fmt.Sprintf("--v=%s", k.LogLevel),
-		"--enable-profiling=false",
-	}
+
+	go k.runServer()
 
 	if k.ClusterConfig.Spec.API.ExternalAddress != "" {
-		serverID, err := util.MachineID()
-		if err != nil {
-			logrus.Errorf("failed to fetch server ID for %v", serviceName)
-		}
-		args = append(args, fmt.Sprintf("--server-count=%d", k.serverCount))
-		args = append(args, fmt.Sprintf("--server-id=%s", serverID))
+		go k.runLeaseCounter()
+	} else {
+		// It's a buffered channel so once we start the runServer routine it'll pick this up and just sees it never changing
+		k.serverCountChan <- 1
 	}
 
-	logrus.Info("Starting konnectivity")
-	k.supervisor = supervisor.Supervisor{
-		Name:    "konnectivity",
-		BinPath: assets.BinPath("konnectivity-server", k.K0sVars.BinDir),
-		DataDir: k.K0sVars.DataDir,
-		RunDir:  k.K0sVars.RunDir,
-		Args:    args,
-		UID:     k.uid,
-	}
+	return k.writeKonnectivityAgent()
+}
 
-	err := k.supervisor.Supervise()
+func (k *Konnectivity) defaultArgs() util.MappedArgs {
+	serverID, err := util.MachineID()
 	if err != nil {
-		return err
+		logrus.Errorf("failed to fetch server ID for konnectivity-server")
```

Comment:
if MachineID() fails here, the serverID will be invalid. not sure if we do anything about it.

Reply:
hmm, dunno really what we could do about it? I think the only reason why it could fail is that `os.Hostname()` fails, but the chances for that are pretty slim.

Reply:
we may want make MachineID() not return an error. But i think thats out of scope here.